internal/pal/pal.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// Package pal provides functions that need to operate differently across different
+// operating systems and/or platforms.
+package pal
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// FilenameInUserHomeDotDirectory returns the full path and filename
+// to the filename in the dotDirectory (e.g. .sqlcmd) in the user's home directory
+// e.g. c:\users\username
+func FilenameInUserHomeDotDirectory(dotDirectory string, filename string) string {
+	home, err := os.UserHomeDir()
+	checkErr(err)
+
+	return filepath.Join(home, dotDirectory, filename)
+}
+
+// UserName returns the name of the currently logged-in user
+func UserName() (userName string) {
+	return username()
+}
+
+// CmdLineWithEnvVars generates a command-line that can be run at the
+// operating system command-line (e.g. bash or cmd) which also requires
+// one or more environment variables to also be set
+func CmdLineWithEnvVars(vars []string, cmd string) string {
+	var sb strings.Builder
+	for _, v := range vars {
+		sb.WriteString(envVarCommand())
+		sb.WriteString(cliQuoteIdentifier() + v + cliQuoteIdentifier())
```

Comment:
is there really value in splitting these up instead of having a single function that takes the variable name? 
This function could accept a map instead of an array, and pass the keys as the var names and the values as the var values. The pal function that exports the environment variable would parse the values and do any escaping of special characters.

Reply:
The value is code coverage reporting :smile: .  I'd like to promote a culture that a lot of the files in this repo are at 100% code coverage ("most" already files are),  (I want to add a "100PercentClub" build pipeline check so a list (or source file metadata tag?) of all files that are at 100% code coverage to never drop below 100%).

There is 2nd reason, the other approach is to use runtime.GOOS == "windows" etc., but this generates a style failure in go vet (and IDEs), because it is flagged as "unreachable" so you never get the little "green tick" for the source file being 100% good.

Reply:
I've added issue:

[https://github.com/microsoft/go-sqlcmd/issues/172](https://github.com/microsoft/go-sqlcmd/issues/172)

Protect source files already at 100% code coverage - add "100PercentClub" build pipeline check to

Reply:
I've added comment to:

[https://github.com/microsoft/go-sqlcmd/issues/171](https://github.com/microsoft/go-sqlcmd/issues/171)

"Another topic around coding guidelines around Platform Abstraction, should we use runtime.GOOS == "windows" etc. or the filename_windows, filename_darwin, filename_linux approach:"

Reply:
I'll refactor func to take a map in next PR

Reply:
I like using the `file_linux` (or build tags) approach so per-platform code isn't intermixed. 
I wasn't asking to merge all the functions into one using the `GOOS=="windows"` check, I just meant to merge the `envVarCommand()` and `cliQuoteIdentifier()` concatenations into a single function. That function would take both the name and the value of the variable so it can properly escape the value for the current platform.